### PR TITLE
task(storage/child-trie): implement store/load of child trie from DB

### PR DESCRIPTION
**Pull request by kishansagathiya**
## Changes

- implement store/load of child trie from DB

## Tests

```

```

## Issues

- #1490

## Primary Reviewer

- @qdm12

### CODE REVIEW

File: lib/trie/child_storage.go [hunk 1]
```diff
@@ -41,27 +43,27 @@ func (t *Trie) GetChild(keyToChild []byte) (*Trie, error) {
...
 	}
 
 	child.Put(key, value)
 	childHash, err := child.Hash()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	t.childTries[origChildHash] = nil
```
Comment (noot) [inline lib/trie/child_storage.go:63]:
```suggestion
	delete(t.childTries, origChildHash)
```

File: lib/trie/child_storage.go [hunk 2]
```diff
@@ -12,6 +12,8 @@ import (
 // ChildStorageKeyPrefix is the prefix for all child storage keys
 var ChildStorageKeyPrefix = []byte(":child_storage:default:")
 
+const ErrChildTrieDoesNotExists = "child trie does not exist at key"
```
Comment (noot) [inline lib/trie/child_storage.go:15]:
```suggestion
const ErrChildTrieDoesNotExist = "child trie does not exist at key"
```

also, this isn't an error type but just a string, do we need to export it or call it `Err`?

File: lib/trie/database.go [hunk 3]
```diff
@@ -23,6 +23,21 @@ var ErrEmptyProof = errors.New("proof slice empty")
...
+			// a new hash, we store at the previous hash.
+			// TODO: delete the hash entry instead of setting it to nil?
+			continue
+		}
+		if err := v.Store(db); err != nil {
+			return err
+		}
+	}
+
+	fmt.Println("stored all child tries successfully")
```
Comment (noot) [inline lib/trie/database.go:39]: remove

File: lib/trie/database.go [hunk 4]
```diff
@@ -23,6 +23,21 @@ var ErrEmptyProof = errors.New("proof slice empty")
 // and the value is the encoded node.
 // Generally, this will only be used for the genesis trie.
 func (t *Trie) Store(db chaindb.Database) error {
+	for _, v := range t.childTries {
+		if v == nil {
+			// there was a child trie with this hash earlier, but
+			// it has changed since then. When storing child trie at
+			// a new hash, we store at the previous hash.
+			// TODO: delete the hash entry instead of setting it to nil?
+			continue
```
Comment (noot) [inline lib/trie/database.go:32]: this shouldn't happen, the entry should be deleted from the map

File: lib/trie/database.go [hunk 5]
```diff
@@ -179,6 +194,21 @@ func (t *Trie) load(db chaindb.Database, curr Node) error {
...
+		err := childTrie.Load(db, common.NewHash(l.Value))
+		if err != nil {
+			// TODO: Avoid only when state.ErrTrieDoesNotExist
+		} else {
+			if err = t.PutChild(l.Value, childTrie); err != nil {
+				return err
+			}
+		}
+
+	}
```
Comment (noot) [inline lib/trie/database.go:210]: we shouldn't do this for all leaf nodes, but we should instead be checking the sub-trie that has the key `ChildStorageKeyPrefix`, as all leaf nodes that contain `ChildStorageKeyPrefix` in them will point to a child trie

File: lib/trie/database.go [hunk 6]
```diff
@@ -348,6 +378,18 @@ func (t *Trie) writeDirty(db chaindb.Batch, curr Node) error {
...
+	if ok {
+		// see if there is a child trie with this hash
+		childTrieHash := common.NewHash(l.Value)
+		childTrie, childTrieexists := t.childTries[childTrieHash]
+		if childTrieexists {
+			if err := childTrie.writeDirty(db, childTrie.root); err != nil {
+				return err
+			}
+		}
+	}
```
Comment (noot) [inline lib/trie/database.go:391]: same here - only check subtrie with `ChildStorageKeyPrefix`

Comment (noot) [reply lib/trie/database.go:391]: alternatively can't you just iterate througn the `childTries` map and call `writeDirty` on them?

Comment (kishansagathiya) [reply lib/trie/database.go]: addressed

File: lib/trie/database.go (hunk 5 above)
Comment (kishansagathiya) [reply lib/trie/database.go]: addressed

File: lib/trie/database.go (hunk 4 above)
Comment (kishansagathiya) [reply lib/trie/database.go]: addressed

File: lib/trie/child_storage.go [hunk 7]
```diff
@@ -41,27 +43,27 @@ func (t *Trie) GetChild(keyToChild []byte) (*Trie, error) {
...
-		return err
+		return nil, err
 	}
 
-	t.childTries[origChildHash] = nil
+	delete(t.childTries, origChildHash)
 	t.childTries[childHash] = child
 
-	return t.PutChild(keyToChild, child)
+	return &childHash, t.PutChild(keyToChild, child)
```
Comment (qdm12) [inline lib/trie/child_storage.go:66]: Why are we returning a point to the hash? It's just an array of 32 bytes

File: lib/trie/child_storage.go [hunk 8]
```diff
@@ -12,6 +12,8 @@ import (
 // ChildStorageKeyPrefix is the prefix for all child storage keys
 var ChildStorageKeyPrefix = []byte(":child_storage:default:")
 
+const ErrChildTrieDoesNotExist = "child trie does not exist at key"
```
Comment (qdm12) [inline lib/trie/child_storage.go:15]: Change that to an exported global error

```suggestion
var ErrChildTrieDoesNotExist = errors.New("child trie does not exist")
```

File: lib/trie/child_storage.go [hunk 9]
```diff
@@ -73,7 +75,7 @@ func (t *Trie) GetFromChild(keyToChild, key []byte) ([]byte, error) {
 	}
 
 	if child == nil {
-		return nil, fmt.Errorf("child trie does not exist at key %s%s", ChildStorageKeyPrefix, keyToChild)
+		return nil, fmt.Errorf("%s %s%s", ErrChildTrieDoesNotExist, ChildStorageKeyPrefix, keyToChild)
```
Comment (qdm12) [inline lib/trie/child_storage.go:78]:
```suggestion
		return nil, fmt.Errorf("%w: at key 0x%x%x", ErrChildTrieDoesNotExist, ChildStorageKeyPrefix, keyToChild)
```

File: lib/trie/child_storage.go [hunk 10]
```diff
@@ -93,7 +95,7 @@ func (t *Trie) ClearFromChild(keyToChild, key []byte) error {
 		return err
 	}
 	if child == nil {
-		return fmt.Errorf("child trie does not exist at key %s%s", ChildStorageKeyPrefix, keyToChild)
+		return fmt.Errorf("%s %s%s", ErrChildTrieDoesNotExist, ChildStorageKeyPrefix, keyToChild)
```
Comment (qdm12) [inline lib/trie/child_storage.go:98]:
```suggestion
		return fmt.Errorf("%w: at key 0x%x%x", ErrChildTrieDoesNotExist, ChildStorageKeyPrefix, keyToChild)
```

File: lib/trie/database.go [hunk 11]
```diff
@@ -23,6 +23,12 @@ var ErrEmptyProof = errors.New("proof slice empty")
 // and the value is the encoded node.
 // Generally, this will only be used for the genesis trie.
 func (t *Trie) Store(db chaindb.Database) error {
+	for _, v := range t.childTries {
+		if err := v.Store(db); err != nil {
+			return err
```
Comment (qdm12) [inline lib/trie/database.go:28]: Please wrap the error with some context information such as the child trie root hash (key in t.childTries)

File: lib/trie/database.go [hunk 12]
```diff
@@ -179,6 +185,19 @@ func (t *Trie) load(db chaindb.Database, curr Node) error {
 		}
 	}
 
+	for _, key := range t.GetKeysWithPrefix([]byte(ChildStorageKeyPrefix)) {
```
Comment (qdm12) [inline lib/trie/database.go:188]:
```suggestion
	for _, key := range t.GetKeysWithPrefix(ChildStorageKeyPrefix) {
```

File: lib/trie/database.go [hunk 13]
```diff
@@ -179,6 +185,19 @@ func (t *Trie) load(db chaindb.Database, curr Node) error {
...
+		childTrie := NewEmptyTrie()
+		value := t.Get(key)
+		err := childTrie.Load(db, common.NewHash(value))
+		if err != nil {
+			// TODO: Avoid only when state.ErrTrieDoesNotExist
```
Comment (qdm12) [inline lib/trie/database.go:193]: What do you mean by `Avoid`? :thinking:

File: lib/trie/database.go [hunk 14]
```diff
@@ -179,6 +185,19 @@ func (t *Trie) load(db chaindb.Database, curr Node) error {
...
+		} else {
+			if err = t.PutChild(value, childTrie); err != nil {
+				return err
```
Comment (qdm12) [inline lib/trie/database.go:196]: Please wrap the error and add context such as the child trie hash

File: dot/state/storage_test.go [hunk 15]
```diff
@@ -183,3 +187,57 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+	ts, err := runtime.NewTrieState(genTrie)
+	require.NoError(t, err)
+
+	header, err := types.NewHeader(bs.GenesisHash(), ts.MustRoot(),
+		common.Hash{}, big.NewInt(1), types.NewDigest())
+	require.NoError(t, err)
+
+	err = storage.StoreTrie(ts, header)
+	require.NoError(t, err)
+	time.Sleep(time.Millisecond * 1000)
```
Comment (qdm12) [inline dot/state/storage_test.go:220]: Why are we sleeping here?

File: lib/trie/child_storage.go [hunk 16]
```diff
@@ -41,27 +43,27 @@ func (t *Trie) GetChild(keyToChild []byte) (*Trie, error) {
 }
 
 // PutIntoChild puts a key-value pair into the child trie located in the main trie at key :child_storage:[keyToChild]
-func (t *Trie) PutIntoChild(keyToChild, key, value []byte) error {
+func (t *Trie) PutIntoChild(keyToChild, key, value []byte) (*common.Hash, error) {
```
Comment (qdm12) [inline lib/trie/child_storage.go:46]: Why are returning the root hash of the child trie? It doesn't seem to be used anywhere right?

**Review by qdm12 — CHANGES_REQUESTED** [review]

**Comment by kishansagathiya** [comment]
I should have made this draft.
There was some code leftover, which I used to test things out, but forgot to remove while making the PR.

Some of the changes that I made earlier were not required after applying changes recommended by noot. 

@qdm12 Made changes you asked for in last two commits.

File: lib/trie/database.go [hunk 17]
```diff
@@ -179,6 +185,21 @@ func (t *Trie) load(db chaindb.Database, curr Node) error {
...
+		if err != nil {
+			return fmt.Errorf("failed to load child trie with root hash=0x%x: %w", value, err)
+		}
+
+		err = t.PutChild(value, childTrie)
+		if err != nil {
+			return fmt.Errorf("failed to insert child trie with root hash=0x%x into main trie: %w",
+				childTrie.root.GetHash(), err)
+		}
+	}
```
Comment (qdm12) [inline lib/trie/database.go:201]: nit could we have this in its own method like `loadChildTries` and call it here?

File: dot/state/storage_test.go [hunk 18]
```diff
@@ -183,3 +187,56 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+
+	_, err = storage.GetStorageChild(&root, []byte("keyToChild"))
+	require.NoError(t, err)
+
+	// Clear trie from cache and fetch data from disk.
+	storage.tries.Delete(root)
+	time.Sleep(time.Millisecond * 100)
+
+	// Should these databases really be different?
+	fmt.Println(storage.db == db)
```
Comment (qdm12) [inline dot/state/storage_test.go:232]: Remove or assert

File: dot/state/storage_test.go [hunk 19]
```diff
@@ -183,3 +187,56 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+
+	root, err := genTrie.Hash()
+	require.NoError(t, err)
+
+	_, err = storage.GetStorageChild(&root, []byte("keyToChild"))
+	require.NoError(t, err)
+
+	// Clear trie from cache and fetch data from disk.
+	storage.tries.Delete(root)
+	time.Sleep(time.Millisecond * 100)
```
Comment (qdm12) [inline dot/state/storage_test.go:229]: Why do we need to sleep here? Aren't the calls blocking?

File: dot/state/storage_test.go [hunk 20]
```diff
@@ -183,3 +187,56 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+	fmt.Println(storage.db == db)
+
+	_, err = storage.GetStorageChild(&root, []byte("keyToChild"))
+	require.NoError(t, err)
+
+	value, err := storage.GetStorageFromChild(&root, []byte("keyToChild"), []byte("keyInsidechild"))
+	require.NoError(t, err)
+
+	require.Equal(t, []byte("voila"), value)
+
```
Comment (qdm12) [inline dot/state/storage_test.go:241]:
```suggestion
```

**Review by qdm12 — APPROVED** [review]
Nice :+1: Sorry I didn't know it was still a drafty PR :wink: Note you can convert it back to draft even if you open it as non-draft (although people will still be notified for reviews).

File: lib/trie/child_storage.go (hunk 2 above)
Comment (kishansagathiya) [reply lib/trie/child_storage.go]: changed it to error type in later commits

File: dot/state/storage_test.go (hunk 18 above)
Comment (kishansagathiya) [reply dot/state/storage_test.go]: removed. (now I know they are deliberately different)

File: lib/trie/database.go (hunk 17 above)
Comment (kishansagathiya) [reply lib/trie/database.go:201]: generally you split when you want to reuse the code or make a large piece of code more readable. This function is not that big. Separating this logic, would not achieve much.

Comment (qdm12) [reply lib/trie/database.go:201]: Yes but also it's easier to test smaller bits. I don't think there is a test covering this specifically for example right (there might a test for a function calling that one though)?

File: dot/state/storage_test.go [hunk 21]
```diff
@@ -183,3 +186,51 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+	require.NoError(t, err)
+
+	header, err := types.NewHeader(blockState.GenesisHash(), ts.MustRoot(),
+		common.Hash{}, big.NewInt(1), types.NewDigest())
+	require.NoError(t, err)
+
+	err = storage.StoreTrie(ts, header)
+	require.NoError(t, err)
+
+	root, err := genTrie.Hash()
```
Comment (qdm12) [inline dot/state/storage_test.go:220]: nit `rootHash` would be better

File: dot/state/storage_test.go [hunk 22]
```diff
@@ -183,3 +186,51 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
...
+	testChildTrie := trie.NewEmptyTrie()
+	testChildTrie.Put([]byte("keyInsidechild"), []byte("voila"))
+
+	err = genTrie.PutChild([]byte("keyToChild"), testChildTrie)
+	require.NoError(t, err)
+
+	storage, err := NewStorageState(db, blockState, genTrie, pruner.Config{})
+	require.NoError(t, err)
+
+	ts, err := runtime.NewTrieState(genTrie)
```
Comment (qdm12) [inline dot/state/storage_test.go:210]: nit `trieState` would be better

**Review by qdm12 — APPROVED** [review]

**Review by EclesioMeloJunior — APPROVED** [review]

File: dot/state/storage_test.go (hunk 21 above)
Comment (kishansagathiya) [reply dot/state/storage_test.go]: addressed

File: dot/state/storage_test.go (hunk 22 above)
Comment (kishansagathiya) [reply dot/state/storage_test.go]: addressed

File: dot/state/storage_test.go [hunk 23]
```diff
@@ -183,3 +187,57 @@ func TestStorage_StoreTrie_NotSyncing(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, 2, syncMapLen(storage.tries))
 }
+
+func TestGetStorageChildAndGetStorageFromChild(t *testing.T) {
+	// initialise database using data directory
+	basepath := "/tmp/gossamer-test-db"
```
[the remark this answers is not in the document]
Comment (qdm12) [reply dot/state/storage_test.go]: Will be done in #2137

**Review by qdm12 — DISMISSED** [review]

**Review by noot — APPROVED** [review]
looks good to me! awesome work :D

File: lib/trie/database.go (hunk 17 above)
Comment (kishansagathiya) [reply lib/trie/database.go:201]: Spent some time on this and wasn't obvious how to test this. I have created a separate issue for this and assigned it to myself. https://github.com/ChainSafe/gossamer/issues/2171

**Pull request merged**